File: WWDC2015/WWDC2015/NBBallView.m
#import "NBBallView.h"
#import "NBItemObject.h"

@interface NBBallView ()

@property (nonatomic, strong) NBItemObject *item;

@end

@implementation NBBallView

- (instancetype)initWithFrame:(CGRect)frame item:(NBItemObject *)item {
    self = [super initWithFrame:frame];
    if (self) {
        self.item = item;
        self.layer.cornerRadius = NBBallViewWidth/2;
        self.backgroundColor = self.item.color;
        self.layer.masksToBounds = YES;
        [self addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapped:)]];
        UIImageView *imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0,
                                                                               self.frame.size.width,
                                                                               self.frame.size.height)];
        [self addSubview:imageView];
        if (self.item.imageName) {
            imageView.image = [UIImage imageNamed:self.item.imageName];
        }
    }
    return self;
}

- (void)tapped:(id)sender {
    [self.delegate ballView:self
  wasSelectedWithItemObject:self.item];
}

@end

File: WWDC2015/NBBallView.swift
import Foundation
import UIKit

@objc protocol NBBallViewDelegate {
    func ballViewWasSelected(ballView: NBBallView, itemObject: NBItemObject)
}

@objc class NBBallView: UIView {
    var item: NBItemObject
    var delegate: NBBallViewDelegate
    
    required init(coder aDecoder: NSCoder) {
        fatalError("This class does not support NSCoding.")
    }
    
    init(frame: CGRect, item: NBItemObject, delegate: NBBallViewDelegate) {
        self.item = item
        self.delegate = delegate;
        super.init(frame: frame)
        self.layer.cornerRadius = 40.0
        self.backgroundColor = item.color
        self.layer.masksToBounds = true
        self.addGestureRecognizer(UITapGestureRecognizer(target: self, action: "tapped:"))
        if (self.item.imageName.length > 0) {
            var imageView = UIImageView(frame: CGRect(
                x: 0,
                y: 0,
                width: self.frame.size.width,
                height: self.frame.size.width));
            self.addSubview(imageView)
            imageView.image = UIImage(named: self.item.imageName as String!)
        }
    }
    
    func tapped(sender: AnyObject) {
        self.delegate.ballViewWasSelected(self, itemObject: self.item)
    }
}